### ISSUE TITLE
Async value between view and model 

### ISSUE THREAD

**Issue by pschwarzer-tt** [angular/angular#48629]
### Which @angular/* package(s) are the source of the bug?

core

### Is this a regression?

Yes

### Description

We found an issue, when modifying a value of an `<input>` using data binding ` <input #quantity [value]="val" (input)="onModelChanged(quantity)" />` or `<input [(ngModel)]="value" />` after user has insert a value.

We want replace one character ( , by . ) during typing with:

```
val = "";

onModelChanged(element: HTMLInputElement) {
    let val = element.value;
    val = val.replace(',', '.');
    this.val = val; // model has corrected value, but view won't update
 }
```
When typing 1,1 into the input, the value correctly get 1.1
But when selecting the . in 1.1 and replacing it with a , the value 1,1 stays in the input, the member this.val gets the corrected value of 1.1 but the view did not recognize this and still has the 1,1 displayed. Using `ChangeDetectorRef ` `markForCheck()` or `detectChanges()` won't help here.

This is only one simple example. We see this issue in other situations too. For example, when we insert a number 1.11. When number than is modified to 1.111 and been rounded in the Java backend to precision 2 and the same value is returned (1.11) than we have the same issue. (input shows 1.111, model value is 1.11)

### Please provide a link to a minimal reproduction of the bug

https://stackblitz.com/edit/angular-fwanqj?file=README.md

### Please provide the exception or error you saw

```true
HTML shows different value than bound by typescript file.
```

### Please provide the environment you discovered this bug in (run `ng version`)

```true
see prepared stackblitz example.
```

**Comment by pkozlowski-opensource** [comment]
This is a super-interesting corner case. Essentially it is the artefacts of how the change detection works: https://stackblitz.com/edit/angular-standalone-m4z1w3?file=src%2Fmain.ts 

The NgModel implementation relies on the standard input bindings.

I would say that the long term solution here is to implement #3009

**Comment by alxhub** [comment]
Yep, this is (quite confusingly) working as intended. Consider `this.val` (which Angular has cached as the "old value" of the `[value]="val"` binding) as this change happens:

* `this.val` starts off as `'1.1'`, as does the `<input>`
* You select the `.` and type a `,`
* `onModelChanged` runs, and reads `'1,1'` from the input
* `onModelChanged` replaces the `,` with a `.` and sets `this.val` to `'1.1'` (the same value as before)
* Angular change detection runs and checks the `[value]="val"` binding
* It compares the previous known value `'1.1'` with the new value `'1.1'` and finds no difference
* It skips updating the binding as a result, leaving the user-typed value of `1,1` in the `<input>`

A way around this is to write `this.val` back into the `<input>` directly in `onModelChanged`. I believe this is internally what `NgModel` is doing to keep its internal model in sync with the `<input>`.

**Comment by pschwarzer-tt** [comment]
Setting the input value directly is how we currently handle this issue. We check if `element.value` is different to `val` and if it is, we set `element.value` manually. 
```
if (!isEqual(val, element.value)) {
      element.value = val; 
}
 ````
But with this fix we got further problems, like setting cursor at right position (after the .). We prefer, that Angular handle this case correctly by itself.
 
 And as explained above, we have this problem also, when rounding values in our Java backend.
